libs/vm_monitor/src/cgroup.rs

```diff
@@ -0,0 +1,689 @@
+
+                        let memory_high =
+                            self.get_high_bytes().context("failed to get memory.high")?;
+                        let new_high = memory_high + self.config.memory_high_increase_by_bytes;
+                        info!(
+                            current_high_bytes = memory_high,
+                            new_high_bytes = new_high,
+                            "updating memory.high"
+                        );
+                        self.set_high_bytes(new_high)
+                            .context("failed to set memory.high")?;
+                        wait_to_increase_memory_high
+                            .as_mut()
+                            .reset(Instant::now() + self.config.memory_high_increase_every)
+                    }
+
+                    // we can't do anything
+                }
+            };
+        }
+    }
+
+    /// Handle a `memory.high`, returning whether we are still waiting on upscale
+    /// by the time the function returns.
+    ///
+    /// The general plan for handling a `memory.high` event is as follows:
+    /// 1. Freeze the cgroup
+    /// 2. Start a timer for `self.config.max_upscale_wait`
+    /// 3. Request upscale
+    /// 4. After the timer elapses or we receive upscale, thaw the cgroup.
+    /// 5. Return whether or not we are still waiting for upscale. If we are,
+    ///    we'll increase the cgroups memory.high to avoid getting oom killed
+    #[tracing::instrument(skip(self))]
+    pub async fn handle_memory_high_event(
+        &self,
+        upscales: &mut Receiver<Sequenced<Resources>>,
+    ) -> anyhow::Result<bool> {
+        // Immediately freeze the cgroup before doing anything else.
+        info!("received memory.high event -> freezing cgroup");
+        self.freeze().context("failed to freeze cgroup")?;
+
+        // We'll use this for logging durations
+        let start_time = Instant::now();
+
+        // Start a timer for the maximum time we'll leave the cgroup frozen for:
+        let thaw_deadline = tokio::time::sleep(self.config.max_upscale_wait);
+
+        // Request the upscale
+        info!(
+            wait = ?self.config.max_upscale_wait,
+            "sending request for immediate upscaling",
+        );
+        self.upscale_requester
+            .send(())
+            .await
+            .context("failed to request upscale")?;
+
+        let mut waiting_on_upscale = true;
+        tokio::select! {
+            // Deadline passed without us receiving upscale
+            _ = thaw_deadline => {
+                info!(elapsed = ?start_time.elapsed(), "timed out waiting for upscale");
+            }
+
+            // Upscaled :)
+            res = self.await_upscale(upscales) => {
+                res.context("failed to await upscale")?;
+                info!(elapsed = ?start_time.elapsed(), "received upscale in time");
+                waiting_on_upscale = false;
+            }
+        }
```

Comment:
This pattern could also be 

```rust
// way earlier where you "start" the sleep
let timeouted = tokio::time::timeout(self.config.max_upscale_wait, self.await_upscale(upscales));

// request

// after requesting it
let waiting_on_upscale = match timeouted.await {
  Ok(res) => {
    res.contextfailed to await upscale")?;
    info!(...);
    false
  }
  Err(_timeout) => {
    info!(...);
    true
  }
};
```

(I know because I was writing quite a few of these which were replaced with `tokio::time::timeout`).

But perhaps no need to start ripping these out now, it just allows to avoid `tokio::select` which is not a trivial macro to expand.

Reply:
Resolved by: 762b4a7570f1b0fa838c03ca34219d46d0922330